File: apps/nongyu-go-track-server/internal/ingest/validate_test.go
package ingest

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestValidateOne_RejectsBadType(t *testing.T) {
	_, err := ValidateOne(RawEvent{EventID: "e1", EventType: "nope", EventName: "x"})
	if err == nil || err.Code != "INVALID_TYPE" {
		t.Fatalf("got %+v", err)
	}
}

func TestValidateOne_TruncatesHugeProps(t *testing.T) {
	s := strings.Repeat("a", 5000)
	raw, err := json.Marshal(map[string]any{"message": s})
	if err != nil {
		t.Fatal(err)
	}
	fields, ierr := ValidateOne(RawEvent{
		EventID:   "e1",
		EventType: "crash",
		EventName: "js",
		Props:     raw,
	})
	if ierr != nil {
		t.Fatal(ierr)
	}
	if !strings.Contains(fields.PropsJSON, `"_truncated":true`) && !strings.Contains(fields.PropsJSON, `"_truncated": true`) {
		t.Fatalf("expected truncated marker, got %s", fields.PropsJSON)
	}
}

func TestValidateOne_OK(t *testing.T) {
	fields, err := ValidateOne(RawEvent{
		EventID:   "550e8400-e29b-41d4-a716-446655440000",
		EventType: "screen_view",
		EventName: "home",
		Platform:  "android",
	})
	if err != nil {
		t.Fatal(err)
	}
	if fields.EventName != "home" {
		t.Fatalf("%+v", fields)
	}
}

File: apps/nongyu-node-track-server/src/ingest/validate.test.ts
import { describe, expect, it } from "vitest";
import { validateOne } from "../ingest/validate.js";

describe("validateOne", () => {
  it("rejects bad type", () => {
    const r = validateOne({ event_id: "e1", event_type: "nope", event_name: "x" });
    expect(r.error?.code).toBe("INVALID_TYPE");
  });

  it("truncates huge props", () => {
    const s = "a".repeat(5000);
    const r = validateOne({
      event_id: "e1",
      event_type: "crash",
      event_name: "js",
      props: { message: s },
    });
    expect(r.error).toBeUndefined();
    expect(r.fields!.propsJson).toContain('"_truncated":true');
  });

  it("accepts ok event", () => {
    const r = validateOne({
      event_id: "550e8400-e29b-41d4-a716-446655440000",
      event_type: "screen_view",
      event_name: "home",
      platform: "android",
    });
    expect(r.error).toBeUndefined();
    expect(r.fields!.eventName).toBe("home");
  });

  it("accepts llm_proxy_fail", () => {
    const r = validateOne({
      event_id: "550e8400-e29b-41d4-a716-446655440001",
      event_type: "llm_proxy_fail",
      event_name: "50210",
    });
    expect(r.error).toBeUndefined();
    expect(r.fields!.eventType).toBe("llm_proxy_fail");
  });
});
